File: src/main/java/io/github/kingg22/api_vacunas_panama/persistence/repository/SedeRepository.java
package io.github.kingg22.api_vacunas_panama.persistence.repository;

import io.github.kingg22.api_vacunas_panama.persistence.entity.Sede;
import io.github.kingg22.api_vacunas_panama.web.dto.UUIDNombreDto;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface SedeRepository extends JpaRepository<Sede, UUID> {

    Optional<Sede> findByNombre(String nombre);

    @Query("SELECT new io.github.kingg22.api_vacunas_panama.web.dto.UUIDNombreDto(s.id, s.nombre) FROM Sede s WHERE s.estado LIKE 'ACTIVO'")
    List<UUIDNombreDto> findAllIdAndNombre();

}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/sede/repository/SedeRepository.kt
package io.github.kingg22.api.vacunas.panama.modules.sede.repository

import io.github.kingg22.api.vacunas.panama.modules.common.dto.UUIDNombreDto
import io.github.kingg22.api.vacunas.panama.modules.sede.entity.Sede
import io.github.kingg22.api.vacunas.panama.util.withSession
import io.quarkus.hibernate.reactive.panache.kotlin.PanacheRepositoryBase
import io.smallrye.mutiny.coroutines.awaitSuspending
import jakarta.enterprise.context.ApplicationScoped
import java.util.UUID

@ApplicationScoped
class SedeRepository : PanacheRepositoryBase<Sede, UUID> {
    suspend fun findByIdOrNull(id: UUID): Sede? = withSession { findById(id).awaitSuspending() }

    suspend fun findAllIdAndNombre(): List<UUIDNombreDto> = withSession { _ ->
        list(
            """
            SELECT s
            FROM Sede s
            JOIN FETCH s.entidad e
            WHERE s.entidad.estado LIKE 'ACTIVO'
            ORDER BY s.entidad.nombre
            """,
        ).awaitSuspending().map { UUIDNombreDto(it.id!!, it.entidad.nombre) }
    }
}
